aten/src/ATen/native/CPUFallback.cpp

```diff
@@ -0,0 +1,141 @@
+      (*stack)[arguments_begin + idx] = std::move(cpu_ivalue);
+    }
+  }
+  auto cpu_tensors = to_cpu(tensor_args);
+
+  for (auto i = 0; i < tensor_args_indices.size(); ++i) {
+    //auto cpu_ivalue = c10::IValue(cpu_tensors[i]);
+    auto idx = tensor_args_indices[i];
+    //(*stack)[arguments_begin + idx] = std::move(cpu_ivalue);
+    (*stack)[arguments_begin + idx] = c10::IValue(cpu_tensors[i]);
+  }
+
+  // Step 2: Call the underlying CPU implementation of the operator
+  op.redispatchBoxed(c10::DispatchKeySet(c10::DispatchKey::CPU), stack);
+
+  // Step 3: We need to take special care to handle mutable aliases properly:
+  // If any input tensors are mutable aliases, we need to
+  // directly copy the updated data on the CPU tensors back to the original inputs.
+  for (int64_t i = 0; i < tensor_args_indices.size(); ++i) {
+    auto tensor_idx = tensor_args_indices[i];
+    const auto& alias_info = schema_args[tensor_idx].alias_info();
+    if (alias_info.has_value() && alias_info.value().isWrite()) {
+      at::_copy_from_and_resize(cpu_tensors[i], tensor_args[i]);
+    }
+  }
+
+  // Step 4: Convert any CPU output tensors back to the original input device.
+  // For mutable alias'd outputs, we also need to take special care
+  // to move the ORIGINAL input tensor back onto the stack, in place of
+  // the temporary CPU output tensor that we created.
+  //
+  // Also note that we are incapable of handling immutable alises properly.
+  // Why?
+  // Schemas with an immutable alias'd tensor outputs correspond to view operators.
+  // For example, the `view_as` schema from native_functions.yaml:
+  // `view_as(Tensor(a) self, Tensor other) -> Tensor(a)`
+  // We can't handle these ops properly, because view ops are supposed to return
+  // a NEW tensor that shares the SAME storage as the original tensor.
+  // However, the new tensor that we created cannot share the same storage,
+  // since it lives on CPU and the original tensor lives on a different device.
+  // Because of that, we treat immutable aliases the same way that we treat non-aliases:
+  // as a fresh tensor that has entirely new storage.
+  const auto& schema_returns = op.schema().returns();
+  const auto& num_returns = schema_returns.size();
+  auto returns = torch::jit::last(stack, num_returns);
+  const auto returns_begin = stack->size() - num_returns;
+
+  for (int64_t idx = 0; idx < returns.size(); ++idx) {
+    if (returns[idx].isTensor()) {
+      const auto& return_tens = returns[idx].toTensor();
+      if (return_tens.defined()) {
+        const auto& alias_info = schema_returns[idx].alias_info();
+        if (alias_info.has_value() && alias_info.value().isWrite()) {
+          // mutable alias case: move the input ivalue directly onto the stack
+          // in place of the existing cpu output tensor.
+          bool found_alias = false;
+          for (int64_t i = 0; i < tensor_args_indices.size(); ++i) {
```

Comment:
🚨 quadratic police 🚨

Reply:
Yeah, this is totally something we could precompute. Right now we have a way to go from `Argument`->`AliasInfo`, but there's no direct link from `AliasInfo` -> [list_of_args_with_that_alias]. I could avoid the runtime quadratic by pre-computing that map in this fallback, but idk if it's worth the effort - it's not reaaaally quadratic since you only do the linear search for each mutable alias'd output argument (there's usually only one). And in xla's case, afaik most people want to call functional ops that don't have any mutable alias'd outputs.

On the topic of boxed fallback perf: there are probably a bunch of things we could pre-compute at registration time to speed this up, and boxed fallbacks in general. They would all increase the size of `FunctionSchema` though, so it'd probably be worth benchmarking to make sure we actually get a speedup.

Just a few ideas (not sure if these are all fruitful):
- [the one you said] Precompute the mapping between alias'd inputs/outputs, which would avoid the double for loop here. Right now you can identify them by checking if their alias_info objects are the same, but we could store direct references between the input/output args instead. I'm not sure whether or not we'd have to deal with weird schemas though, like `(Tensor(a!) t1, Tensor(a!) t2) -> Tensor(a!)`
- Precompute which arguments are tensors/tensorlists, since it's probably a pretty common task in a boxed fallback to iterate through tensor arguments.
- Also I see the FunctionSchema stores a `std::vector<Argument>` - it might be an easy win to make these `SmallVector`.